### PR TITLE
Add Metal V1 embed pooling support

### PR DESCRIPTION
**Pull request by CAICAIIs**
Implements the first Metal V1 pooling MVP for #361: text-only `embed` support on Apple Silicon.

What this adds
- Advertises `embed` for compatible text-only pooling models on Metal V1
- Returns `ModelRunnerOutput.pooler_output` for pooling requests
- Keeps `sampled_token_ids` empty for embedding requests
- Pools the final prompt-token hidden state with LAST pooling and L2 normalization
- Fails fast for unsupported variants such as classify, score, token-level pooling, non-LAST sequence pooling, multimodal inputs, and prompt embeddings

What is intentionally out of scope
- Direct model-provided embedding tensors
- Classification / rerank / late interaction
- Multimodal embeddings
- Token-level pooling
- Sequence pooling strategies other than LAST (`MEAN`, `CLS`, `ALL`, `STEP`)

Validation
- `python -m pytest tests/test_v1_pooling.py -q`
- `python -m pytest tests/test_v1_worker.py tests/test_v1_pooling.py -q`
- `python -m pytest tests/test_v1_pooling.py tests/test_v1_model_runner_generate.py -q`
- `python -m pytest tests/test_v1_model_runner_generate.py tests/test_v1_worker.py tests/test_paged_attention.py tests/test_sampling.py tests/test_grammar_bitmask.py -q`
- `scripts/lint.sh`
- Offline smoke: `LLM.embed()` with `mlx-community/Qwen3-Embedding-0.6B-8bit` returned 2 embeddings of 1024 dims
- Online smoke: `POST /v1/embeddings` on `vllm serve` returned 2 embeddings of 1024 dims

### CODE REVIEW

File: docs/supported_models.md [hunk 1]
```diff
@@ -11,6 +11,41 @@ vllm-metal currently focuses on text-only language models on Apple Silicon. Mult
...
+The Metal runner rejects classification, reranking, late interaction,
+token-level pooling, multimodal embeddings, scheduled encoder inputs, prompt
+embeddings, and unsafe dimension requests with diagnostic errors. Do not add a
+model row to the table below until a real `LLM.embed` or `/v1/embeddings`
+smoke passes on Apple Silicon with the model name, revision, command, and output
+dimension recorded.
+
+Verified embedding smoke:
+
+| Model | Support | Runner | Evidence | Notes |
```
Comment (WindChimeRan) [inline docs/supported_models.md:20]: This small embedding model table should stay here. But others (explanations, how to use .etc) should be in a separated feature md.

Comment (CAICAIIs) [reply docs/supported_models.md:20]: done

File: vllm_metal/v1/worker.py
Comment (WindChimeRan) [inline vllm_metal/v1/worker.py:1]: Compatible for what? Dual path is adding maintainance burden.

File: vllm_metal/v1/model_runner.py [hunk 2]
```diff
@@ -899,11 +1032,67 @@ def _start_paged_forward(
             scheduler_output=scheduler_output,
             logits=logits,
             target_hidden_states=target_hidden_states,
+            pooling_hidden_states=pooling_hidden_states,
             cu_seqlens=cu_seqlens,
             decode_segments=decode_segments,
             num_decode_tokens=num_decode_tokens,
         )
 
+    def _finish_paged_pooling_batch(
```
Comment (WindChimeRan) [inline vllm_metal/v1/model_runner.py]: Can you simplify the code  by 
1. move more code from model_runner to pooling, with higher abstraction? most of pooling details should not go into model_runner. 
2. I understand this PR is a MVP thus many other poolings are NotImplementedError, but please try to not writing long list of if - else
3. modifying other scopes of functionality needs explanation.

**Review by WindChimeRan — COMMENTED** [review]
Please document explicitly which pooling strategies are supported (LAST) vs not yet implemented (MEAN, CLS, ALL, STEP)  so future contributors know what to add next without reading the gate logic.

File: docs/text_embedding_pooling.md
Comment (WindChimeRan) [inline docs/text_embedding_pooling.md:1]: link it to [mkdocs.yaml](https://github.com/vllm-project/vllm-metal/blob/main/mkdocs.yaml)

**Review by WindChimeRan — COMMENTED** [review]
Direct embedding seems to be a speculative implementation and not full functional? If so, we should drop it.

**Review by WindChimeRan — COMMENTED** [review]
I saw many patterns of over complicated engineering & over abstraction  in this PR. 

Please only keep high value unit tests. 

Please simplify the code, and focus on narrower scope.

**Comment by WindChimeRan** [comment]
@CAICAIIs please check the CI test fails

**Comment by CAICAIIs** [comment]
done, thx your review

File: vllm_metal/v1/model_runner.py [hunk 3]
```diff
@@ -526,11 +547,25 @@ def profile_run(self) -> int:
         mx.clear_cache()
         cache_before = mx.get_cache_memory()
         dummy_tokens = mx.zeros((1, warmup_len), dtype=mx.int32)
-        mx.eval(self._extract_logits(self._forward_model(dummy_tokens)))
+        mx.eval(*self._dummy_forward_outputs(dummy_tokens))
         overhead = mx.get_cache_memory() - cache_before
         mx.set_cache_limit(overhead)
         return overhead
 
+    def _dummy_forward_outputs(self, input_ids: mx.array) -> list[mx.array]:
```
Comment (WindChimeRan) [inline vllm_metal/v1/model_runner.py:604]: Try to use the same patter with `stt`. don't wanna balloon model runner. 

Eventually, both, stt and pooling should be abstract out of the model runner

**Review by WindChimeRan — COMMENTED** [review]
Simplify the code: I meant less code and fewer concepts. The feature is only supporting one type of pooling. The implementation should be about that size.  Please delete the speculative parts (dense-head probe, signature inspection, the tests covering them). 

try to decouple pooling and model_runner: less pooling code in model_runner.

**Comment by CAICAIIs** [comment]
> Simplify the code: I meant less code and fewer concepts. The feature is only supporting one type of pooling. The implementation should be about that size. Please delete the speculative parts (dense-head probe, signature inspection, the tests covering them).
> 
> try to decouple pooling and model_runner: less pooling code in model_runner.

Sorry, I misunderstood your point earlier and made the first version too broad. I’ve corrected it now: removed the speculative dense-head/signature code and kept this PR focused on the narrow text-only embed pooling path. Thanks for the clarification.

**Comment by ricky-chaoju** [comment]
@CAICAIIs Please resolve the conflict.

**Comment by CAICAIIs** [comment]
> @CAICAIIs Please resolve the conflict.

done

**Review by WindChimeRan — COMMENTED** [review]
You haven't resolved previous round of review. especially on the model_runner hot path

*  `_finish_paged_pooling_batch` should not live on the runner.
*  Deduplicate `getattr(self.model_config, "runner_type", None) == "pooling"`
— follow the `_is_stt` pattern.

**Comment by CAICAIIs** [comment]
> You haven't resolved previous round of review. especially on the model_runner hot path
> 
> * `_finish_paged_pooling_batch` should not live on the runner.
> * Deduplicate `getattr(self.model_config, "runner_type", None) == "pooling"`
>   — follow the `_is_stt` pattern.

Sorry, I missed part of your previous review around the model_runner hot path.

I’ve updated the PR to move the paged pooling finish logic out of model_runner, deduplicate the pooling runner check with an _is_pooling flag following the _is_stt pattern, and keep the scope focused on text-only LAST embed pooling.

I also reran the targeted tests, lint, and real e2e validation with both LLM.embed and /v1/embeddings.

**Review by WindChimeRan — APPROVED** [review]
Thanks!
